src/rpp/rpp/operators/fwd/first.hpp

```diff
@@ -0,0 +1,78 @@
+//                  ReactivePlusPlus library
+//
+//          Copyright Aleksey Loginov 2022 - present.
+//                            TC Wang 2022 - present.
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          https://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/victimsnino/ReactivePlusPlus
+//
+
+#pragma once
+
+#include <rpp/observables/constraints.hpp>
+#include <rpp/observables/details/member_overload.hpp>
+#include <rpp/utils/function_traits.hpp>
+
+#include <string>
+
+namespace rpp::details
+{
+struct first_tag;
+}
+
+namespace rpp::details
+{
+
+struct unmatched_emission_count_exception
+{
+    std::string what{"expect one emission before completed"};
+};
```

Comment:
No, no, no. Any exceptions in C++ should be inherited from `std::runtime_error` in case of something wrong during runtime and `std::logic_error` if something unexpected happens. Also, the name of exception doesn't make sense. So, in your case it should be something like this
```suggestion
struct not_enough_emissions : std::runtime_error
{
    using std::runtime_error::runtime_error;
};
```
You can place it in something like `rpp/utils/exceptions.hpp` and use there. During throwing you can add details about exception something like "the first() operator requires at least one emission before complete"

Reply:
Oh side question, I found some articles saying we could throw any types.

Reply:
Yeah, it is true, C++ provides ability to throw anything as exception. But also it it better to make exception in tree-like way for better handling and understanding. For example, as I’ve mentioned before, there are two types of error: logic and runtime. And in most cases if we want to “catch all exceptions” we catch runtime because logic means something absolutely unexpected and application can’t continue to work. Same can be propagated to another thing: we can create our subtype “RPP_exception” and inherit all our exceptions from this one. So, in case of we/user want to catch only RPP exceptions we can easily do it